### DISCUSSION TITLE
msvc怎么关闭特定的警告？

### DISCUSSION THREAD

**Discussion by coderling** [xmake-io/xmake#3034] [Q&A] [answered]
开启了
set_warnings("all", "error")
后续关闭特定警告

`rule("cm-cxflags")
    on_config(function (target)
        local cx_flags = {
            "-Wno-unused-parameter", 
            "-Wno-unused-command-line-argument"
        }
        -- clang cl 情况下是可行的
        if target:has_tool("cxx", "clang_cl")  then
            target:add("cxxflags", cx_flags)
        end 
        if target:has_tool("cc", "clang_cl") then
            target:add("cflags", cx_flags)
        end
        
        -- msvc 但在msvc下，编译的时候会报命令行错误
        if target:has_tool("cxx", "cl")  then
            target:add("cxxflags", "/Wd4819", {public = true, force = true})
        end 
        if target:has_tool("cc", "cl") then
            target:add("cflags", "/Wd4819", {public = true, force = true})
        end
    end)
rule_end()`

error: cl : Command line error D8021 : invalid numeric argument '/Wd4819'

**Comment by waruqi** [comment, marked as answer]
它不是提示了么，'/Wd4819 没这个参数，你强加进去 当然会报错，删了

**Reply by coderling** [reply]
E:\GitProj\LPrimitive\src\allocator\interface\Allocator.hpp(1): error C2220: the following warning is treated as an error
E:\GitProj\LPrimitive\src\allocator\interface\Allocator.hpp(1): warning C4819: The file contains a character that cannot be represented in the current code page (936). Save the file in Unicode format to prevent
data loss
E:\GitProj\LPrimitive\src\utils\interface\Align.hpp(1): warning C4819: The file contains a character that cannot be represented in the current code page (936). Save the file in Unicode format to prevent data loss

那我要怎么关掉特定warning，在cmake中可以通过target_compile_options()

**Reply by waruqi** [reply]
> 在cmake中可以通过target_compile_options()

add_cflags,add_cxflags, add_cxxflags 。。。

刚都说了，cl.exe 报错，它都明显提示了 说没这个 flag，跟 xmake 有啥关系，为啥还要去研究 xmake 的接口呢？

怎么就不去分析下 为啥没这个 flag，这不很明了么，就一种可能，你传的 flag 不对，然后去 msdn 那边仔细看下 flag 说明不就好了。。

要学会分析问题。。


https://learn.microsoft.com/en-us/cpp/build/reference/compiler-option-warning-level?view=msvc-170

> /wdnnnn | Suppresses the compiler warning that is specified by nnnn.For example, 
> /wd4326 suppresses compiler warning C4326.


仔细看，大小写。


你现在传的值，即使放到cmake的target_compile_options 一样报错

**Reply by coderling** [reply]
的确，没注意到大小写，非常感谢回复
